common/metrics/go-metrics-m3.go

```diff
@@ -0,0 +1,159 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package metrics
+
+import (
+	"fmt"
+	"time"
+
+	gometrics "github.com/rcrowley/go-metrics"
+	"github.com/uber-common/bark"
+)
+
+const (
+	delayToFirstLogDump  = time.Minute * 15
+	delayBetweenLogDumps = time.Minute * 60
+	delayBetweenExports  = time.Second * 10
+)
+
+// goMetricsExporter allows an rcrowley/go-metrics registry to be
+// periodically exported to the Cherami metrics system
+type goMetricsExporter struct {
+	c     Client
+	scope int
+
+	// metricNameToMetric maps a metric name to the metric symbol,
+	// e.g. "request-rate" -> OutputhostCGKafkaRequestRate
+	metricNameToMetric  map[string]int
+	closeCh             <-chan struct{}
+	metricNameToLastVal map[string]int64
+	l                   bark.Logger
+	registry            gometrics.Registry // Registry to be exported
+	lastLogDump         time.Time
+}
+
+// NewGoMetricsExporter starts the exporter loop to export go-metrics to the Cherami
+// metrics system and returns the go-metrics registry that should be used
+func NewGoMetricsExporter(
+	c Client,
+	scope int,
+	metricNameToMetric map[string]int,
+	l bark.Logger,
+	closeCh <-chan struct{},
+) gometrics.Registry {
+	// Make a new registy and give it to the caller. This allows the exporters to operate independently
+	registry := gometrics.NewRegistry()
+	r := &goMetricsExporter{
+		c:                   c,
+		scope:               scope,
+		registry:            registry,
+		metricNameToMetric:  metricNameToMetric,
+		l:                   l,
+		closeCh:             closeCh,
+		metricNameToLastVal: make(map[string]int64),
+
+		// Calculate the time to provide the delayToFirstLogDump in export()
+		lastLogDump: time.Now().Add(-1 * delayBetweenLogDumps).Add(delayToFirstLogDump),
+	}
+	go r.run()
+	return registry
+}
+
+func (r *goMetricsExporter) run() {
+	t := time.NewTicker(delayBetweenExports)
+	defer t.Stop()
+
+exportLoop:
+	for {
+		r.export()
+
+		select {
+		case <-t.C:
+		case <-r.closeCh:
+			break exportLoop
+		}
+	}
+}
+
+func (r *goMetricsExporter) export() {
+	r.registry.Each(func(name string, i interface{}) {
```

Comment:
The general problem here is that any metrics solution has a pattern of local-reporting->aggregation->remote-reporting, but rcrowley metrics and m3 both want to do the aggregation step. So I have to de-aggregate from rcrowley to allow m3 to aggregate.

Regarding the Each() call, this is just saying that I get a stab at handling all metrics that Sarama reports through rcrowley/go-metrics. I don't handle all of them. The unhandled ones are periodically (but minimally) reported in the unexported metrics section. I include that unexported reporting, since there may come a time where those metrics are useful for some kind of debugging. Right now, it's a bunch of broker-specific metrics, which I fear will exceed the cardinality limits that M3 imposes on us. See https://godoc.org/github.com/Shopify/sarama

For the handled metrics, I need to de-aggregate them in various ways, and then map them to the equivalent M3 stuff.